api/envoy/service/status/v2/csds.proto

```diff
@@ -0,0 +1,92 @@
+syntax = "proto3";
+
+package envoy.service.status.v2;
+
+import "envoy/admin/v2alpha/config_dump.proto";
+import "envoy/api/v2/core/base.proto";
+import "envoy/type/matcher/metadata.proto";
+import "envoy/type/matcher/string.proto";
+
+import "google/api/annotations.proto";
+import "google/protobuf/struct.proto";
+
+option java_package = "io.envoyproxy.envoy.service.status.v2";
+option java_outer_classname = "CsdsProto";
+option java_multiple_files = true;
+option java_generic_services = true;
+
+// CSDS is Client Status Discovery Service. It can be used to get the status of
+// an xDS-compliant client from the management server's point of view. In the
+// future, it can potentially be used as an interface to get the current
+// state directly from the client.
+service ClientStatusDiscoveryService {
+  rpc StreamClientStatus(stream ClientStatusRequest) returns (stream ClientStatusResponse) {
+  }
+
+  rpc FetchClientStatus(ClientStatusRequest) returns (ClientStatusResponse) {
+    option (google.api.http).post = "/v2/discovery:client_status";
+    option (google.api.http).body = "*";
+  }
+}
+
+// Status of a config.
+enum ConfigStatus {
+  // Status info is not available/unknown.
+  UNKNOWN = 0;
+
+  // Management server has sent the config to client and received ACK.
+  SYNCED = 1;
+
+  // Config is not sent.
+  NOT_SENT = 2;
+
+  // Management server has sent the config to client but hasn’t received
+  // ACK/NACK.
+  STALE = 3;
+
+  // Management server has sent the config to client but received NACK.
+  ERROR = 4;
+}
+
+// Request for client status of clients identified by a list of NodeMatchers.
+message ClientStatusRequest {
+  message NodeMatcher {
+    // Specifies match criteria on the client node id.
+    type.matcher.StringMatcher node_id = 1;
+
+    // Specifies match criteria on the client node metadata.
+    type.matcher.MetadataMatcher node_metadata = 2;
+  }
+
+  // Management server can use these match criteria to identify clients.
+  repeated NodeMatcher node_matchers = 1;
+}
+
+// Detailed config (per xDS) with status.
+// [#next-free-field: 6]
+message PerXdsConfig {
+  ConfigStatus status = 1;
+
+  oneof per_xds_config {
+    admin.v2alpha.ListenersConfigDump listener_config = 2;
+
+    admin.v2alpha.ClustersConfigDump cluster_config = 3;
+
+    admin.v2alpha.RoutesConfigDump route_config = 4;
+
+    admin.v2alpha.ScopedRoutesConfigDump scoped_route_config = 5;
```

Comment:
Do these need to be Anys in the case of mixed v2/v3, similar to the problem that @htuch just faced on config dump? cc @htuch

Reply:
Nup, the `*ConfigDump` protos already have the `Any` messages inside them.

Reply:
:+1: